kite-data/kite-data-core/src/main/java/org/kitesdk/data/spi/filesystem/SignalManager.java

```diff
@@ -0,0 +1,194 @@
+import org.kitesdk.data.DatasetException;
+import org.kitesdk.data.DatasetIOException;
+import org.kitesdk.data.spi.Constraints;
+
+import com.google.common.base.Joiner;
+
+/**
+ * Manager for creating, and checking {@link View#isReady() ready} signals.
+ * Stored in a filesystem, typically HDFS.
+ */
+public class SignalManager {
+
+  private final Path signalDirectory;
+  private final FileSystem rootFileSystem;
+
+  private static final String UNBOUNDED_CONSTRAINT = "unbounded";
+
+  /**
+   * Creates a new signal manager using the given signal directory.
+   *
+   * @param conf the Hadoop configuration
+   * @param signalDirectory directory in which the manager
+   *                        stores signals.
+   *
+   * @return a signal manager instance.
+   */
+  public static SignalManager create(Configuration conf, Path signalDirectory) {
+    try {
+      FileSystem rootFileSystem = signalDirectory.getFileSystem(conf);
+
+      rootFileSystem.mkdirs(signalDirectory);
+
+      return new SignalManager(signalDirectory, rootFileSystem);
+    } catch (IOException e) {
+      throw new DatasetIOException("Unable to create signal manager directory: "
+              + signalDirectory, e);
+    }
+  }
+
+  /**
+   * Loads a signal manager that stores metadata data under the given signal
+   * metadata directory if it exists. Returns <code>null</code> if it does not.
+   *
+   * @param conf the Hadoop configuration
+   * @param signalDirectory directory in which the manager stores signals.
+   *
+   * @return a signal manager instance, or <code>null</code> if the given
+   * directory does not exist.
+   */
+  public static SignalManager load(Configuration conf, Path signalDirectory) {
+    try {
+      FileSystem rootFileSystem = signalDirectory.getFileSystem(conf);
+
+      if (rootFileSystem.exists(signalDirectory)) {
+        return new SignalManager(signalDirectory, rootFileSystem);
+      } else {
+        return null;
+      }
+    } catch (IOException e) {
+      throw new DatasetIOException ("Cannot load signal manager at:"
+              + signalDirectory, e);
+    }
+  }
+
+  private SignalManager(Path signalDirectory, FileSystem rootFileSystem)
+          throws IOException  {
+    this.signalDirectory = signalDirectory;
+    this.rootFileSystem = rootFileSystem;
+  }
+
+  /**
+   * Create a signal for the specified constraints.
+   *
+   * @param viewConstraints The constraints to create a signal for.
+   *
+   * @throws DatasetException if the signal could not be created.
+   */
+  public void signalReady(Constraints viewConstraints) {
+    String normalizedConstraints = getNormalizedConstraints(viewConstraints);
+
+    Path signalPath = new Path(signalDirectory, normalizedConstraints);
+    try{
+      // create the output stream to overwrite the current contents, if the directory or file
+      // exists it will be overwritten to get a new timestamp
+      FSDataOutputStream os = rootFileSystem.create(signalPath, true);
```

Comment:
I think it would be valuable to be able to signal a path that already has been by using [`setTimes`](https://hadoop.apache.org/docs/r2.2.0/api/org/apache/hadoop/fs/FileSystem.html#setTimes(org.apache.hadoop.fs.Path, long, long)) if the file already exists. Then you'd be able to use signals to reprocess if new data comes in.